dipy/align/imaffine.py

```diff
@@ -0,0 +1,777 @@
+                                        self.moving_vals,
+                                        self.samples[..., :self.dim],
+                                        mgrad)
+
+        # Evaluate the mutual information and its gradient
+        # The results are in self.metric_val and self.metric_grad
+        # ready to be returned from 'distance' and 'gradient'
+        self.update_mi_metric(update_gradient)
+
+    def distance(self, xopt):
+        r""" Numeric value of the negative Mutual Information
+
+        We need to change the sign so we can use standard minimization
+        algorithms.
+
+        Parameters
+        ----------
+        xopt : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+
+        Returns
+        -------
+        val : float
+            the negative mutual information of the input images after warping
+            the moving image by the currently set transform with `xopt`
+            parameters
+        """
+        if self.samples is None:
+            self._update_dense(xopt, False)
+        else:
+            self._update_sparse(xopt, False)
+        return -1 * self.metric_val
+
+    def gradient(self, xopt):
+        r""" Numeric value of the metric's gradient at the given parameters
+
+        Parameters
+        ----------
+        xopt : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+
+        Returns
+        -------
+        grad : array, shape(n,)
+            the gradient of the negative Mutual Information
+        """
+        if self.samples is None:
+            self._update_dense(xopt, True)
+        else:
+            self._update_dense(xopt, True)
+        return self.metric_grad * (-1)
+
+    def value_and_gradient(self, xopt):
+        r""" Numeric value of the metric and its gradient at given parameters
+
+        Parameters
+        ----------
+        xopt : array, shape (n,)
+            the parameter vector of the transform currently used by the metric
+            (the transform name is provided when self.setup is called), n is
+            the number of parameters of the transform
+
+        Returns
+        -------
+        val : float
+            the negative mutual information of the input images after warping
+            the moving image by the currently set transform with `xopt`
+            parameters
+        grad : array, shape(n,)
+            the gradient of the negative Mutual Information
+        """
+        if self.samples is None:
+            self._update_dense(xopt, True)
+        else:
+            self._update_sparse(xopt, True)
+        return -1 * self.metric_val, self.metric_grad * (-1)
```

Comment:
> Reason for different -1 uses here?

Too much time dealing with custom operators in c++ (I guess my subconscious triggers a warning whenever I put an operator in front of an object)... but yeah, it looks like a mess and there's no reason for being inconsistent. Changed to -1 * x for both.